### Case study  - Uber Data Analysis

The data of a driver’s uber trips are available for year 2016.
Your manager wants you to explore this data to give him some useful insights about the trip behaviour of a Uber driver.

#### Dataset - 
The dataset contains Start Date, End Date, Start Location, End Location, Miles Driven and Purpose of drive (Business, Personal, Meals, Errands, Meetings, Customer Support etc.)


#### Steps - 

1.Import the libraries

2.Get the data and observe it

3.Check missing values, either remove it or fill it.

4.Get summary of data using python function.

5.Explore the data parameter wise

Here we have information of destination(start and stop), time(start and stop), category and purpose of trip, miles covered.


In [ ]:
# ----------------------
# Concepts To cover 
# ----------------------
# 1. Data profiling
# 2. group by function
# 3. Apply function 
# 4. DateTime operations 

In [1]:
# Import the libraries 
import numpy as np
import pandas as pd

In [2]:
# Read the Data 
df = pd.read_csv('My Uber Drives - 2016.csv')

In [3]:
# View first n rows of data 
# n = 7
df.head(7)

START_DATE*       END_DATE* CATEGORY*           START*            STOP*  \
0  1/1/2016 21:11  1/1/2016 21:17  Business      Fort Pierce      Fort Pierce   
1   1/2/2016 1:25   1/2/2016 1:37  Business      Fort Pierce      Fort Pierce   
2  1/2/2016 20:25  1/2/2016 20:38  Business      Fort Pierce      Fort Pierce   
3  1/5/2016 17:31  1/5/2016 17:45  Business      Fort Pierce      Fort Pierce   
4  1/6/2016 14:42  1/6/2016 15:49  Business      Fort Pierce  West Palm Beach   
5  1/6/2016 17:15  1/6/2016 17:19  Business  West Palm Beach  West Palm Beach   
6  1/6/2016 17:30  1/6/2016 17:35  Business  West Palm Beach       Palm Beach   

   MILES*         PURPOSE*  
0     5.1   Meal/Entertain  
1     5.0              NaN  
2     4.8  Errand/Supplies  
3     4.7          Meeting  
4    63.7   Customer Visit  
5     4.3   Meal/Entertain  
6     7.1          Meeting

In [4]:
# View the last 5 rows of data
df.tail()

START_DATE*         END_DATE* CATEGORY*            START*  \
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                 STOP*   MILES*        PURPOSE*  
1151  Unknown Location      3.9  Temporary Site  
1152  Unknown Location     16.2         Meeting  
1153           Gampaha      6.4  Temporary Site  
1154         Ilukwatta     48.2  Temporary Site  
1155               NaN  12204.7             NaN

In [5]:
# Understand shape of data 
print(df.shape)

(1156, 7)


In [6]:
# There are 1156 rows and 7 columns in the dataset.

In [7]:
# Understand shape of data 
print(df.size)

8092


In [8]:
# size = total number of elements placed in 1D array.

In [9]:
# Same as above, gives non-null number of records
df.isnull().sum()

START_DATE*      0
END_DATE*        1
CATEGORY*        1
START*           1
STOP*            1
MILES*           0
PURPOSE*       503
dtype: int64

In [10]:
# Non-Null values
df.notnull().sum()

START_DATE*    1156
END_DATE*      1155
CATEGORY*      1155
START*         1155
STOP*          1155
MILES*         1156
PURPOSE*        653
dtype: int64

1. PURPOSE column has lots of missing values  
2. 1155 or 1156 records ??

In [11]:
df[df['PURPOSE*'].isnull()]

START_DATE*         END_DATE* CATEGORY*            START*  \
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
32      1/19/2016 9:09    1/19/2016 9:23  Business       Whitebridge   
85      2/9/2016 10:54    2/9/2016 11:07  Personal       Whitebridge   
86      2/9/2016 11:43    2/9/2016 11:50  Personal        Northwoods   
87      2/9/2016 13:36    2/9/2016 13:52  Personal        Tanglewood   
...                ...               ...       ...               ...   
1066  12/19/2016 14:37  12/19/2016 14:50  Business  Unknown Location   
1069  12/19/2016 19:05  12/19/2016 19:17  Business         Islamabad   
1071   12/20/2016 8:49   12/20/2016 9:24  Business  Unknown Location   
1143  12/29/2016 20:53  12/29/2016 21:42  Business           Kar?chi   
1155            Totals               NaN       NaN               NaN   

                    STOP*   MILES* PURPOSE*  
1             Fort Pierce      5.0      NaN  
32    Lake Wellingborough      7.2      NaN  
85             Northwoods      5.3      NaN  
86             Tanglewood      3.0      NaN  
87                Preston      5.1      NaN  
...                   ...      ...      ...  
1066     Unknown Location      5.4      NaN  
1069     Unknown Location      2.2      NaN  
1071           Rawalpindi     12.0      NaN  
1143     Unknown Location      6.4      NaN  
1155                  NaN  12204.7      NaN  

[503 rows x 7 columns]

Show the records with missing values for column= PURPOSE 

In [12]:
df[~df['PURPOSE*'].isnull()]

START_DATE*         END_DATE* CATEGORY*            START*  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
5       1/6/2016 17:15    1/6/2016 17:19  Business   West Palm Beach   
...                ...               ...       ...               ...   
1150   12/31/2016 1:07   12/31/2016 1:14  Business           Kar?chi   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   

                 STOP*  MILES*         PURPOSE*  
0          Fort Pierce     5.1   Meal/Entertain  
2          Fort Pierce     4.8  Errand/Supplies  
3          Fort Pierce     4.7          Meeting  
4      West Palm Beach    63.7   Customer Visit  
5      West Palm Beach     4.3   Meal/Entertain  
...                ...     ...              ...  
1150           Kar?chi     0.7          Meeting  
1151  Unknown Location     3.9   Temporary Site  
1152  Unknown Location    16.2          Meeting  
1153           Gampaha     6.4   Temporary Site  
1154         Ilukwatta    48.2   Temporary Site  

[653 rows x 7 columns]

How many records are non-null  / have values ( in a particular column )

In [13]:
print(df[~df['PURPOSE*'].isnull()].shape[0])

653


### Renaming columns

In [14]:
# Rename the columns to remove the * from the names
df.rename(columns = {'START_DATE*' : 'START_DATE',
                    'END_DATE*' : 'END_DATE',
                    'CATEGORY*' : 'CATEGORY',
                    'START*' : 'START',
                    'STOP*' : 'STOP',
                    'MILES*' : 'MILES',
                    'PURPOSE*' : 'PURPOSE'},
                     inplace = True)

#or 

# Replace the * character from all the columns.
#df.columns = df.columns.str.replace('*', '')

# Using lambda function
#df.rename(mapper=lambda x: x.replace('*', ''), axis=1, inplace=True)
#df.rename(columns = lambda x : str(x)[:-1])

In [15]:
df.head()

START_DATE        END_DATE  CATEGORY        START             STOP  \
0  1/1/2016 21:11  1/1/2016 21:17  Business  Fort Pierce      Fort Pierce   
1   1/2/2016 1:25   1/2/2016 1:37  Business  Fort Pierce      Fort Pierce   
2  1/2/2016 20:25  1/2/2016 20:38  Business  Fort Pierce      Fort Pierce   
3  1/5/2016 17:31  1/5/2016 17:45  Business  Fort Pierce      Fort Pierce   
4  1/6/2016 14:42  1/6/2016 15:49  Business  Fort Pierce  West Palm Beach   

   MILES          PURPOSE  
0    5.1   Meal/Entertain  
1    5.0              NaN  
2    4.8  Errand/Supplies  
3    4.7          Meeting  
4   63.7   Customer Visit

### Filtering dataframes -1 

In [16]:
# shows the entries where PURPOSE is null
print(df[df['PURPOSE'].isnull()])
# inverting the selection ( not null ) ( works for booleans cases)
print(df[~df['PURPOSE'].isnull()])

            START_DATE          END_DATE  CATEGORY             START  \
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
32      1/19/2016 9:09    1/19/2016 9:23  Business       Whitebridge   
85      2/9/2016 10:54    2/9/2016 11:07  Personal       Whitebridge   
86      2/9/2016 11:43    2/9/2016 11:50  Personal        Northwoods   
87      2/9/2016 13:36    2/9/2016 13:52  Personal        Tanglewood   
...                ...               ...       ...               ...   
1066  12/19/2016 14:37  12/19/2016 14:50  Business  Unknown Location   
1069  12/19/2016 19:05  12/19/2016 19:17  Business         Islamabad   
1071   12/20/2016 8:49   12/20/2016 9:24  Business  Unknown Location   
1143  12/29/2016 20:53  12/29/2016 21:42  Business           Kar?chi   
1155            Totals               NaN       NaN               NaN   

                     STOP    MILES PURPOSE  
1             Fort Pierce      5.0     NaN  
32    Lake Wellingborough      7.2     NaN  


### Filtering dataframe - 2 

In [17]:
# Conditions within dataframe 
print(df[df['STOP']=='Fort Pierce'])
# Inverting the selection 
print(df[~(df['STOP']=='Fort Pierce')])

       START_DATE        END_DATE  CATEGORY        START         STOP  MILES  \
0  1/1/2016 21:11  1/1/2016 21:17  Business  Fort Pierce  Fort Pierce    5.1   
1   1/2/2016 1:25   1/2/2016 1:37  Business  Fort Pierce  Fort Pierce    5.0   
2  1/2/2016 20:25  1/2/2016 20:38  Business  Fort Pierce  Fort Pierce    4.8   
3  1/5/2016 17:31  1/5/2016 17:45  Business  Fort Pierce  Fort Pierce    4.7   

           PURPOSE  
0   Meal/Entertain  
1              NaN  
2  Errand/Supplies  
3          Meeting  
            START_DATE          END_DATE  CATEGORY             START  \
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
5       1/6/2016 17:15    1/6/2016 17:19  Business   West Palm Beach   
6       1/6/2016 17:30    1/6/2016 17:35  Business   West Palm Beach   
7       1/7/2016 13:27    1/7/2016 13:33  Business              Cary   
8       1/10/2016 8:05    1/10/2016 8:25  Business              Cary   
...                ...               ...       ...            

Explore the details from the MILES column


In [18]:
# Show the top 10 rides (*in terms of distance driven)
print(df.sort_values(by = 'MILES', ascending = False).head(10))

# Show the row that has the max miles 
print(df.sort_values(by = 'MILES', ascending = False).head(1))

            START_DATE          END_DATE  CATEGORY             START  \
1155            Totals               NaN       NaN               NaN   
269    3/25/2016 16:52   3/25/2016 22:22  Business             Latta   
270    3/25/2016 22:54    3/26/2016 1:39  Business      Jacksonville   
881   10/30/2016 15:22  10/30/2016 18:23  Business         Asheville   
776    9/27/2016 21:01    9/28/2016 2:37  Business  Unknown Location   
546    7/14/2016 16:39   7/14/2016 20:05  Business       Morrisville   
559    7/17/2016 12:20   7/17/2016 15:25  Personal             Boone   
297     4/2/2016 19:38    4/2/2016 22:36  Business      Jacksonville   
299      4/3/2016 2:00     4/3/2016 4:16  Business          Florence   
727    8/27/2016 16:15   8/27/2016 19:13  Business  Unknown Location   

                  STOP    MILES         PURPOSE  
1155               NaN  12204.7             NaN  
269       Jacksonville    310.3  Customer Visit  
270          Kissimmee    201.0         Meeting  
881    

In [19]:
df.isnull().sum()

START_DATE      0
END_DATE        1
CATEGORY        1
START           1
STOP            1
MILES           0
PURPOSE       503
dtype: int64

In [20]:
# Or use other filters to remove this row
df = df.dropna(subset = ['STOP'])

In [21]:
# Get a summary of the numerical columns in the data
df.describe()

MILES
count  1155.000000
mean     10.566840
std      21.579106
min       0.500000
25%       2.900000
50%       6.000000
75%      10.400000
max     310.300000

The miles driven ranges from 0.5 miles to 12204 miles with an average of 21 miles

Max miles - looks suspect

#### Dropping rows  which have null values

In [22]:
# Get the initial data with dropping the NA values
df_1 = df.copy()

# Removal of NULL values
df_1 = df_1.dropna()

# Get the shape of the dataframe after removing the null values
print(df_1.shape)

(653, 7)


The filtered dataset with no nulls ( in PURPOSE column )  contains 653 rows of non-null values

### PANDAS PROFILING
Use some packages that will auto profile your data ( than having to manually type in all the commands )

You can find the documentation here ( https://pandas-profiling.github.io/pandas-profiling/docs/ )

Github page here - https://github.com/pandas-profiling/pandas-profiling

In [23]:
import pandas_profiling as pp
profile = pp.ProfileReport(df_1)  # Of you just want to show here 

# output into an html file 
profile.to_file('output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

### Lets explore the data parameter wise - 

1.Destination - (starting and stopping)

2.Time - (hour of the day, day of week, month of year)

3.Categories

4.Purpose 

5.Grouping two parameters to get more insights


## 1. Understanding  the start and stop points 

In [24]:
# Get the unique starting point, unique destination
# names of unique start points
print(df['START'].unique())
print(df['STOP'].unique())

['Fort Pierce' 'West Palm Beach' 'Cary' 'Jamaica' 'New York' 'Elmhurst'
 'Midtown' 'East Harlem' 'Flatiron District' 'Midtown East'
 'Hudson Square' 'Lower Manhattan' "Hell's Kitchen" 'Downtown' 'Gulfton'
 'Houston' 'Eagan Park' 'Morrisville' 'Durham' 'Farmington Woods'
 'Whitebridge' 'Lake Wellingborough' 'Fayetteville Street' 'Raleigh'
 'Hazelwood' 'Fairmont' 'Meredith Townes' 'Apex' 'Chapel Hill'
 'Northwoods' 'Edgehill Farms' 'Tanglewood' 'Preston' 'Eastgate'
 'East Elmhurst' 'Jackson Heights' 'Long Island City' 'Katunayaka'
 'Unknown Location' 'Colombo' 'Nugegoda' 'Islamabad' 'R?walpindi'
 'Noorpur Shahan' 'Heritage Pines' 'Westpark Place' 'Waverly Place'
 'Wayne Ridge' 'Weston' 'East Austin' 'West University' 'South Congress'
 'The Drag' 'Congress Ave District' 'Red River District' 'Georgian Acres'
 'North Austin' 'Coxville' 'Convention Center District' 'Austin' 'Katy'
 'Sharpstown' 'Sugar Land' 'Galveston' 'Port Bolivar' 'Washington Avenue'
 'Briar Meadow' 'Latta' 'Jacksonville'

In [25]:
# count of unique start points using len()
print(len(df['START'].unique()))

177


In [26]:
# count of unique start points using len()
len(pd.unique(df['START']))

177

In [27]:
# or use can use the nunique function
print(df['STOP'].nunique())

188


Stations which are appeared in both start and stop locations 

In [28]:
print(df[df['START'] == df['STOP']][['START', 'STOP']])

                 START              STOP
0          Fort Pierce       Fort Pierce
1          Fort Pierce       Fort Pierce
2          Fort Pierce       Fort Pierce
3          Fort Pierce       Fort Pierce
5      West Palm Beach   West Palm Beach
...                ...               ...
1147           Kar?chi           Kar?chi
1148           Kar?chi           Kar?chi
1149           Kar?chi           Kar?chi
1150           Kar?chi           Kar?chi
1152  Unknown Location  Unknown Location

[288 rows x 2 columns]


In [29]:
print(df[df['START'].isin(df['STOP'])][['START', 'STOP']])

                 START              STOP
0          Fort Pierce       Fort Pierce
1          Fort Pierce       Fort Pierce
2          Fort Pierce       Fort Pierce
3          Fort Pierce       Fort Pierce
4          Fort Pierce   West Palm Beach
...                ...               ...
1149           Kar?chi           Kar?chi
1150           Kar?chi           Kar?chi
1151           Kar?chi  Unknown Location
1152  Unknown Location  Unknown Location
1154           Gampaha         Ilukwatta

[1130 rows x 2 columns]


In [30]:
# Identify popular start points - top 10
df['START'].value_counts().head(10)

Cary                201
Unknown Location    148
Morrisville          85
Whitebridge          68
Islamabad            57
Durham               37
Lahore               36
Raleigh              28
Kar?chi              27
Westpark Place       17
Name: START, dtype: int64

In [31]:
# Identify popular stop destinations - top 10
df['STOP'].value_counts().head(10)

Cary                203
Unknown Location    149
Morrisville          84
Whitebridge          65
Islamabad            58
Durham               36
Lahore               36
Raleigh              29
Kar?chi              26
Apex                 17
Name: STOP, dtype: int64

In [32]:
# Are there cases where the start and the stop location are the same  ? 
print(len(df[df['START'] == df['STOP']][['START', 'STOP']]))

288


In [33]:
# Favorite starting point wrt the total miles covered 
df.groupby(by='START').MILES.sum().sort_values(ascending=False).head(1)

START
Unknown Location    1976.5
Name: MILES, dtype: float64

In [34]:
# Or use other filters to remove this row
df = df.dropna(subset = ['START'])

In [35]:
# Starting Point with highest miles covered
df[df['MILES'] == max(df['MILES'])][['START', 'MILES']]

START  MILES
269  Latta  310.3

In [36]:
df.dtypes

START_DATE     object
END_DATE       object
CATEGORY       object
START          object
STOP           object
MILES         float64
PURPOSE        object
dtype: object

In [37]:
df['START'].value_counts()

Cary                201
Unknown Location    148
Morrisville          85
Whitebridge          68
Islamabad            57
                   ... 
Florence              1
Ridgeland             1
Daytona Beach         1
Sky Lake              1
Gampaha               1
Name: START, Length: 177, dtype: int64

#### Find out most farthest start and stop pair - top10 ( aggregation ) ( BY TOTAL miles COVERED EVER ! )

In [38]:
# Dropping Unknown Location Value  - Save into another dataframe ( you dont want to overwrite the original df)
df_2 = df[df['START'] != 'Unknown Location']
df_2 = df_2[df_2['STOP'] != 'Unknown Location']

The most popular start and stop pair - ( BY COUNT of travels! )

In [39]:
# The most popular start and stop pair - ( BY COUNT of travels!)
df_2.groupby(['START', 'STOP']).size().sort_values(ascending=False).reset_index(name = 'COUNT')

START                       STOP  COUNT
0       Morrisville                       Cary     75
1              Cary                Morrisville     67
2              Cary                       Cary     53
3              Cary                     Durham     36
4            Durham                       Cary     32
..              ...                        ...    ...
338  Hell's Kitchen                    Midtown      1
339       Hazelwood                     Weston      1
340       Hazelwood  Lexington Park at Amberly      1
341      Hayesville                     Topton      1
342   Winston Salem                  Asheville      1

[343 rows x 3 columns]

**The most popular start to destination pair is Morrisville-Cary**

## 2. Manipulating date & time objects

We will be using the datetime module (  https://docs.python.org/3/library/datetime.html  )

Search for "strftime() and strptime() Format Codes" in the documentation page

In [40]:
# Create columns by converting the start and end date into a datatime format
df['START_DATE'] = pd.to_datetime(df['START_DATE'])
df['END_DATE'] = pd.to_datetime(df['END_DATE'])
df.dtypes

START_DATE    datetime64[ns]
END_DATE      datetime64[ns]
CATEGORY              object
START                 object
STOP                  object
MILES                float64
PURPOSE               object
dtype: object

In [41]:
# Which month did he get most drives  ? 
df['Month'] = pd.DatetimeIndex(df['START_DATE']).month
df.head()

START_DATE            END_DATE  CATEGORY        START  \
0 2016-01-01 21:11:00 2016-01-01 21:17:00  Business  Fort Pierce   
1 2016-01-02 01:25:00 2016-01-02 01:37:00  Business  Fort Pierce   
2 2016-01-02 20:25:00 2016-01-02 20:38:00  Business  Fort Pierce   
3 2016-01-05 17:31:00 2016-01-05 17:45:00  Business  Fort Pierce   
4 2016-01-06 14:42:00 2016-01-06 15:49:00  Business  Fort Pierce   

              STOP  MILES          PURPOSE  Month  
0      Fort Pierce    5.1   Meal/Entertain      1  
1      Fort Pierce    5.0              NaN      1  
2      Fort Pierce    4.8  Errand/Supplies      1  
3      Fort Pierce    4.7          Meeting      1  
4  West Palm Beach   63.7   Customer Visit      1

In [42]:
df['Month'].value_counts()

12    146
8     133
11    122
2     115
3     113
7     112
6     108
10    106
1      61
4      54
5      49
9      36
Name: Month, dtype: int64

In [43]:
# Getting the average distance covered each month
df.groupby("Month").mean()

MILES
Month           
1       8.408197
2       7.897391
3      14.990265
4      20.611111
5       7.424490
6       7.803704
7      10.933929
8      10.041353
9      16.716667
10     17.075472
11      6.695902
12      6.721233

In [44]:
# Which day did he get most drives  ? 
df['Day'] = pd.DatetimeIndex(df['START_DATE']).day
df['Day'].value_counts()

19    67
29    51
27    51
26    45
22    45
21    45
12    44
1     43
16    42
18    38
20    38
13    37
11    37
8     37
28    36
10    35
25    35
7     34
23    34
6     33
14    33
4     33
5     33
3     33
17    33
30    32
2     30
24    29
9     29
15    28
31    15
Name: Day, dtype: int64

Now lets try to answer some questions from this data.

Question1: How many miles was earned per category and purpose ?

Question2: What is percentage of business miles vs personal?

Question3: How much time was spend for drives per category and purpose? 


In [45]:
# Question1: How many miles was earned per category and purpose ?
df.groupby(['CATEGORY','PURPOSE']).sum()['MILES'].sort_values(ascending=False).reset_index(name = 'Total Miles')

CATEGORY          PURPOSE  Total Miles
0  Business          Meeting       2851.3
1  Business   Customer Visit       2089.5
2  Business   Meal/Entertain        911.7
3  Business   Temporary Site        523.7
4  Business  Errand/Supplies        508.0
5  Business  Between Offices        197.0
6  Personal          Commute        180.2
7  Personal           Moving         18.2
8  Business   Airport/Travel         16.5
9  Personal      Charity ($)         15.1

In [46]:
# Question2: What is percentage of business miles vs personal?
BUS=df.groupby('CATEGORY').sum().loc['Business']['MILES']
PER=df.groupby('CATEGORY').sum().loc['Personal']['MILES']
print('Business : ',BUS*100/(BUS+PER))
print('Personal : ',PER*100/(BUS+PER))

Business :  94.11947856153776
Personal :  5.880521438462232


In [47]:
import datetime as dt

In [60]:
# Question 3: How much time was spend for drives per category and purpose?
